lfroms/COMP4109-Project
# Bind UI to socket connection from the backend

### What is the problem being solved in this PR?
This PR connects the frontend to the backend, and adds support for sending/receiving messages directly into rooms.

### Related issues or PRs
Resolves #14

### How did you accomplish this and why did you choose this approach?
- By performing various refactors.
- By creating a new `SocketContext` context provider to provide a single socket connection to the entire app.
- By refactoring the socket implementation on the backend so that it better supports multiple concurrent rooms.
- By creating a series of React hooks for connecting to conversations and sending/receiving messages.

### How to test changes
1. `yarn start:dev`
2. Open 4 browser tabs, two pointing to `/chat/1` and two pointing to `/chat/2`.
3. Send messages into both conversations.  Messages should appear in each conversation tab, and should not appear in the other conversation.

### Checklist
